include/sunlinsol/sunlinsol_ginkgobatch.hpp

```diff
@@ -0,0 +1,505 @@
+
+  void SetScalingMode(int scaling_mode = LAGGED_SCALING)
+  {
+    scaling_mode_ = scaling_mode;
+  }
+
+  /// Sets the left and right scaling vectors to be used.
+  void SetScalingVectors(N_Vector s1, N_Vector s2)
+  {
+    SUNFunctionBegin(sunCtx());
+
+    if (s1 && s2)
+    {
+      s1_ = s1;
+      col_scale_vec_ =
+        std::move(impl::WrapBatchScalingArray(GkoExec(), num_batches_, s1_));
+
+      if (!s2inv_.Convert())
+      {
+        s2inv_ = sundials::experimental::NVectorView(N_VClone(s2));
+      }
+
+      // SUNLinearSolver API wants s2inv_
+      N_VInv(s2, s2inv_);
+      SUNCheckLastErrVoid();
+
+      row_scale_vec_ =
+        std::move(impl::WrapBatchScalingArray(GkoExec(), num_batches_, s2inv_));
+
+      scaling_initialized_ = true;
+    }
+    else
+    {
+      scaling_mode_        = NO_SCALING;
+      scaling_initialized_ = false;
+    }
+  }
+
+  int Setup(BatchMatrix<GkoBatchMatType>* A)
+  {
+    if (num_batches_ != A->NumBatches()) { return SUN_ERR_ARG_OUTOFRANGE; }
+
+    matrix_   = A;
+    do_setup_ = true;
+
+    return SUN_SUCCESS;
+  }
+
+  int Solve(N_Vector b, N_Vector x, sunrealtype tol)
+  {
+    SUNLogInfo(sunLogger(), "linear-solver", "solver = ginkgo (batched)");
+
+    // Applying scaling every solve requires us to redo all the Ginkgo solver setup
+    // This is because the scaling happens in memory, and the Ginkgo solver needs
+    // to know about the matrix values when it is constructed.
+    if (scaling_mode_ == SOLVE_SCALING) { do_setup_ = true; }
+
+    if (do_setup_)
+    {
+      // Create the solver factory
+      solver_factory_ = GkoBatchSolverType::build()             //
+                          .with_max_iterations(max_iters_)      //
+                          .with_tolerance(SUN_UNIT_ROUNDOFF)    //
+                          .with_tolerance_type(tolerance_type_) //
+                          .with_preconditioner(precon_factory_) //
+                          .on(GkoExec());
+
+      if (scaling_mode_ == LAGGED_SCALING || scaling_mode_ == SOLVE_SCALING)
+      {
+        // This will scale the matrix in place.
+        // So if a Ginkgo preconditioner is used, it is applied to the scaled matrix:
+        //    \tilde{A} = P_1^{-1} S_1 A S_2^{-1} P_2^{-1},
+        // but in memory, we just have
+        //    \tilde{A} = S_1 A S_2^{-1}.
+        matrix_->GkoMtx()->scale(row_scale_vec_, col_scale_vec_);
+      }
+
+      solver_ = solver_factory_->generate(matrix_->GkoMtx());
+
+      do_setup_ = false;
+    }
+
+    solver_->reset_tolerance(tol);
+
+    if (!logger_)
+    {
+      logger_ =
+        gko::share(gko::batch::log::BatchConvergence<sunrealtype>::create());
+    }
+    solver_->add_logger(logger_);
+
+    std::unique_ptr<GkoBatchVecType> x_vec{
+      impl::WrapBatchVector(GkoExec(), num_batches_, x)};
+    std::unique_ptr<GkoBatchVecType> b_vec{
+      impl::WrapBatchVector(GkoExec(), num_batches_, b)};
+
+    if (scaling_mode_)
```

Comment:
Match the `if` above to not rely on `NO_SCALING` being zero
```suggestion
    if (scaling_mode_ == LAGGED_SCALING || scaling_mode_ == SOLVE_SCALING)
```